clang/docs/checkedc/Update-to-latest-LLVM-sources.md

```diff
@@ -1,81 +1,212 @@
+```
+    git remote add upstream https://github.com/llvm/llvm-project.git
+    git pull upstream <12.0.0-commit-hash>
+```
+3. Push the branch `release_12.x` to the remote repository, run ADO tests
+on it and fix test failures to make sure that the branch is stable.
+```
+    git push origin release_12.x
+```
+## Practical Tips
+
+1. If you use automated testing, make sure to clean the build directory first.
+Enough changes will likely have accumulated that things may go wrong without
+doing that first.
+
+2. The test failures usually stem from incorrect merges or Checked C-specific data
+not being initialized by new or changed constructor methods.
+
+3. You'll want to run automated tests on Linux and Windows x86/x64, as well as
 Linux LNT tests.  You may find in some cases that tests need to be updated
 based on new warnings or slightly different compiler behavior.
 
-## Merge your branched baseline and master branches
-
-Once all testing is passing, you can merge your branches back into
-your baseline and master branches.
-
-
-    git checkout baseline
-    git merge updated-baseline
-    git checkout master
-    git merge updated-master
-
-## Push the updated branches to GitHub
-
-The changes will be extensive enough that you don't want to do a pull request
+4. The changes will be extensive enough that you don't want to do a pull request
 on GitHub.  Just push the branches up to GitHub.
 
+
+## Handling Merge Conflicts and Test Failures
+
+1. When merge.renamelimit is set to a large value, the output of the `git merge`
+command may have several lines like:
+CONFLICT (rename/delete): ... in master renamed to ... in HEAD. Version HEAD ...
+left in tree.
+CONFLICT (rename/delete): ... in master renamed to ... in HEAD. Version master
+... left in tree.
+
+    If the version in HEAD is left in tree, it is very likely a valid change -
+    the file may have been moved. If the version in `master` is left in tree,
+    the files may be valid files that have been added as part of Checked C
+    support:
+    - `llvm/projects/checkedc-wrapper/lit.site.cfg.py`
+    - `llvm/projects/checkedc-wrapper/lit.cfg.py`
+
+    Others may need more investigation. For example, in the LLVM/Clang 11 to 12
+    merge, two files which were reported by git as "Version master ... left in
+    tree" were old versions lying around that needed to be deleted:
+    - `clang/include/clang/StaticAnalyzer/Core/PathSensitive/SMTAPI.h`
+    - `clang/include/clang/AST/TypeNodes.def`
+
+2. While resolving merge conflicts, it is a good idea to have three reference
+code snapshots to perform diffs of individual files to help distinguish between
+checkedc-specific changes and llvm-11-to-12 changes:
+    - The `master` branch of the `checkedc-clang` repository at the commit hash
+  which was merged into `updated_baseline_master_12`.
+    - The pristine LLVM/Clang source at &lt;branch-point-of-11-on-main&gt;
+    - The pristine LLVM/Clang source at &lt;branch-point-of-12-on-main&gt;
+
```

Comment:
I wholeheartedly agree with the value of being able to see all three versions of a file involved in a merge.  However, the description of the three versions here is accurate only for the initial merge from LLVM 11 to 12 in phase 1 step 6.  For a merge of additional Checked C feature development in phase 1 step 8, the relevant versions are `master`, the baseline LLVM 12 source, and _the merge of the previous `master` with the baseline LLVM 12 source_ that was completed in step 6 or an earlier iteration of step 8.

We are already assuming that `git merge` correctly identifies the three file versions to merge for files without conflicts, so for files that have pending conflicts, it is probably best to view the three file versions already identified by Git rather than try to look them up manually (which could be prone to mistakes/misunderstandings).  I know of a few ways to do that, which might be worth documenting in this file:

1. Best: Run `git mergetool`, assuming you have a suitable merge tool installed and configured.  I currently use kdiff3, but I don't know if it works on Windows or if other tools are better.
2. Set `git config merge.conflictstyle diff3` before running `git merge` so that the conflict markup written to the file shows all three versions rather than just two.  I recommend leaving this option on all the time; I've found that the extra information never hurts.
3. Run `git show :1:clang/path/to/some/File.cpp >clang/path/to/some/File.1.cpp` (and similarly with 2 or 3 in place of 1) to write out the three versions for inspection with other tools.